tensorflow_model_optimization/python/core/sparsity_tf2/riglpruner.py

```diff
@@ -0,0 +1,366 @@
+          connection is dropped then regrown
+        grown_mask_reshaped: mask for weights of the grown connections
+        mask: the current iteration's binary mask over weights
+    """
+    if drop_regrow_reinit:
+      # if dropped then regrown, reinitialize to the value
+      new_connections = tf.math.equal(grown_mask_reshaped, 1)
+    else:
+      new_connections = tf.math.logical_and(
+        tf.math.equal(grown_mask_reshaped, 1), tf.math.equal(mask, 0)
+      )
+    return new_connections
+
+  def _get_grow_tensor(self, weight, method, step=0):
+    if method == 'zeros':
+      grow_tensor = tf.zeros_like(weight, dtype=weight.dtype)
+    elif method == 'random_normal':
+      divisor = 1.
+      stdev = tf.math.reduce_std(weight)
+      grow_tensor = self._random_normal(step, weight.get_shape(), stddev=stdev, dtype=weight.dtype, seed=self._seed) / divisor
+    elif method == 'random_uniform':
+      mean = tf.math.reduce_mean(tf.math.abs(weight))
+      divisor = 1.
+      grow_tensor = self._random_uniform(step, weight.get_shape(), minval=-mean, maxval=mean, 
+                                        dtype=weight.dtype, seed=self._seed) / divisor
+    return grow_tensor
+
+  def _generic_top_k(self, scores, mask, n_to_modify, n_total):
+     # sort the entire array since TPU requires k to be constant
+     # so that its aggregation methods can be reliable
+    _, sorted_idx = tf.math.top_k(
+      tf.reshape(scores, [-1]), k=n_total
+    )
+    expanded_sorted_idx = tf.expand_dims(sorted_idx, 1)
+    new_values = tf.where(
+      tf.range(n_total) < n_to_modify,
+      tf.ones_like(sorted_idx, dtype=mask.dtype),
+      tf.zeros_like(sorted_idx, dtype=mask.dtype)
+    )
+    updated_mask = tf.scatter_nd(expanded_sorted_idx, new_values, new_values.shape)
+
+    return updated_mask
+
+  def _update_mask(self, step, update_fraction, mask, weight, grad):
+    """Called by _maybe_update_block_mask.
+    Updates mask based on weight and grad information.
+
+    Args:
+      step: current iteration from optimizer
+      update_fraction: the fraction of existing connections to update
+      mask: the mask to update
+      weight: trainable variable representing weights
+      grad: gradients obtained from the optimizer
+    """
+    # compute the top k magnitudes then update the current mask
+    drop_scores = self._get_drop_weights(mask, weight, noise_std=self._noise_std, step=step)
+    # need access to exactly which entries are growing to zero out optimizer slot
+    grow_scores = self._get_grow_grads(mask, grad)
+    n_total = tf.size(drop_scores)
+    n_ones = tf.cast(tf.reduce_sum(mask), dtype=tf.int32) # floor not ceiling like sparsity
+    n_prune = tf.cast(
+      tf.cast(n_ones, dtype=tf.float32) * update_fraction, tf.int32
+    )
+    n_keep = n_ones - n_prune
+
+    dropped_mask = self._generic_top_k(drop_scores, mask, n_keep, n_total)
+
+    if grow_scores is not None:
```

Comment:
How would `grow_scores` ever be None if it's defined as `math.abs(grads)`?

Also, this block of code here (to build a `grow` mask) is what should be put in a util method, *not* just `math.abs(grads)`.

Reply:
it would be None if we didn't want to grow (i.e. if we jsut want to do pruning and no growing - which would be another method that's a variant of rigl) -- this is would be specified in another flag though for now

Reply:
so i.e. the variant would subclass this rigl pruner and override the `get_grow_grads` function, which could be to return `None` if it doesn't wish to grow.